### PR TITLE
Allow users to extend config_value API

### PR DESCRIPTION
**Pull request by Neverlord**
Allow users to add custom types to CAF's config value API by specializing `config_value_access`.

The new utility class `config_value_object_access` also makes it rather simpler for users to specialize `config_value_access`. Users only need to implement a trait class for `config_value_object_access` to get all the low-level sum type calls as well as the parsing code out of their way.

The center piece of the trait class is a getter for fields of the user-defined type. This new fields-based API either works with pointers to members or with pairs of getter and setter functions. In both cases, users can also add default values and validity predicates.

### CODE REVIEW

File: libcaf_core/test/make_config_value_field.cpp [hunk 1]
```diff
@@ -0,0 +1,261 @@
...
+  CAF_MESSAGE("without foo member");
+  {
+    CAF_REQUIRE(holds_alternative<foobar>(x, "my-value"));
+    CAF_REQUIRE(get_if<foobar>(&x, "my-value") != caf::none);
+    auto fb = get<foobar>(x, "my-value");
+    CAF_CHECK_EQUAL(fb.foo, 123);
+    CAF_CHECK_EQUAL(fb.bar, "hello");
+    CAF_MESSAGE("with foo member");
+  }
+  put(x, "my-value.foo", 42);
```
Comment (josephnoir) [inline libcaf_core/test/make_config_value_field.cpp]: Not that important, but the `CAF_MESSAGE` comments on the put outside of the block. Would make sense to group them together.

File: libcaf_core/test/make_config_value_field.cpp [hunk 2]
```diff
@@ -0,0 +1,261 @@
...
+  {
+    CAF_REQUIRE(holds_alternative<foobar>(x, "my-value"));
+    CAF_REQUIRE(get_if<foobar>(&x, "my-value") != caf::none);
+    auto fb = get<foobar>(x, "my-value");
+    CAF_CHECK_EQUAL(fb.foo, 42);
+    CAF_CHECK_EQUAL(fb.bar, "hello");
+    CAF_MESSAGE("with foo member");
```
Comment (josephnoir) [inline libcaf_core/test/make_config_value_field.cpp]:
```suggestion
```

File: libcaf_core/caf/config_value.hpp [hunk 3]
```diff
@@ -423,6 +477,34 @@ struct select_config_value_access<T, select_config_value_hint::is_list> {
...
+
+    static void parse_cli(string_parser_state& ps, T& xs) {
+      config_value::dictionary result;
+      bool has_open_token = ps.consume('[');
+      do {
+        if (has_open_token && ps.consume(']')) {
+          ps.skip_whitespaces();
+          ps.code = ps.at_end() ? pec::success : pec::trailing_character;
+          return;
+        }
```
[the remark this answers is not in the document]
Comment (Neverlord) [reply libcaf_core/caf/config_value.hpp]: Yes, our config syntax supports trailing comma. I think the CLI syntax should not deviate from that.

File: libcaf_core/caf/config_value_object_access.hpp [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+/// ~~
+template <class Trait>
+struct config_value_object_access {
+  using object_type = typename Trait::object_type;
+
+  static std::string type_name() {
+    return Trait::type_name();
+  }
+
+  static bool extract(const config_value* src, object_type* dst) {
```
[the remark this answers is not in the document]
Comment (Neverlord) [reply libcaf_core/caf/config_value_object_access.hpp]: Yes. We only need it once to the dictionary out via `get_if`.

File: libcaf_core/src/make_config_option.cpp [hunk 5]
```diff
@@ -75,43 +77,34 @@ config_value bool_get_neg(const void* ptr) {
...
+
+template <uint64_t Denominator>
+void store_timespan(void* ptr, const config_value& x) {
+  *static_cast<size_t*>(ptr) = static_cast<size_t>(get<timespan>(x).count())
+                               / Denominator;
+}
+template <uint64_t Denominator>
+config_value get_timespan(const void* ptr) {
+  auto ival = static_cast<int64_t>(*static_cast<const size_t*>(ptr));
+  timespan val{ival * 1000};
```
[the remark this answers is not in the document]
Comment (Neverlord) [reply libcaf_core/src/make_config_option.cpp]: Oops. 🙈

Good catch!

**Pull request merged**